Add BaseWeightsService

A-Artemis opened the pull request:
### Summary

This PR introduces a new `BaseWeightsService` which handles downloading and fetching of model's pretrained weights.

### How to test

<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/open-edge-platform/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/open-edge-platform/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/open-edge-platform/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

A-Artemis committed:
Add BaseWeightsService

A-Artemis committed:
Fix cleanup in base weights service tests to remove the correct pretrained weights path (ignoring errors)

A-Artemis committed:
Updated timeouts in case of large files.

A-Artemis commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,261 @@
+            # Move temporary file to final location
+            temp_path.rename(local_path)
+            logger.info(f"Successfully downloaded and verified weights: {local_path}")
+        except requests.RequestException as e:
+            logger.error(f"Failed to download weights from {remote_url}: {e}")
+        except Exception as e:
+            logger.error(f"Unexpected error downloading weights from {remote_url}: {e}")
```
@leoll2 What do you think the behaviour should be here?

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+    """Service for downloading and managing pretrained model weights from external archives."""
+
+    def __init__(self, data_dir: Path) -> None:
+        self._weights_cache_dir = data_dir / "models" / "pretrained_weights"
```
```suggestion
        self._weights_cache_dir = data_dir / "pretrained_weights"
```
The root `models` folder is going to be removed, instead models will be stored inside the respective project folder (#4755).

The `pretrained_weights` dir is supposed to be directly under `data_dir`: https://github.com/open-edge-platform/training_extensions/blob/6ec7c7d6ae419eca1e2683db1e8db6def3e3939f/docs/geti_tune/models.md?plain=1#L126-L127

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+            ValueError: If the model manifest is not found, task type mismatch, or doesn't have pretrained weights
+        """
+        manifest = SupportedModels.get_model_manifest_by_id(model_manifest_id)
+        if manifest.id == "null":
```
My bad for not pointing out in the previous PR, but I don't like the `NullModelManifest` class. A more Pythonic and maintainable approach is to raise a `ManifestNotFoundException` in `SupportedModels.get_model_manifest_by_id` instead of returning a null-like object.

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+        manifest = SupportedModels.get_model_manifest_by_id(model_manifest_id)
+        if manifest.id == "null":
+            raise ValueError(f"Model manifest '{model_manifest_id}' not found")
+        if manifest.task.lower() != task.name.lower():
```
I suggest to update the `ModelManifest.task` attribute to be `TaskType` instead of string, so you can do a direct comparison. It's also more robust because a ValidationError is raised when loading a manifest whose YAML references an unrecognized task type.

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+        if manifest.task.lower() != task.name.lower():
+            raise ValueError(f"Task mismatch: expected '{task.name.lower()}', got '{manifest.task.lower()}'")
+        if not manifest.pretrained_weights.url or manifest.pretrained_weights.url == "null":
+            raise ValueError(f"No pretrained weights configured for model '{model_manifest_id}'")
```
This type of validation is not responsibility of `BaseWeightsService`. Since `ModelManifest.pretrained_weights.url` is not optional, this check is already performed by Pydantic when loading the manifest from YAML.

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+        manifest = self._get_and_validate_model_manifest(task, model_manifest_id)
+
+        task_dir = self._weights_cache_dir / task.name.lower()
+        task_dir.mkdir(parents=True, exist_ok=True)
```
I suggest to move the logic to create the task directories to the constructor. In other words, in the constructor you should iterate over the task types and create the missing folders. This way we can detect much earlier issues like incorrect write permissions on the data_dir, already when the application is launched.

leoll2 commented on application/backend/app/services/base_weights_service.py:
```diff
@@ -0,0 +1,262 @@
+        Returns:
+            bool: True if the file integrity is valid, False otherwise
+        """
+        if sha_sum == "null" or not sha_sum:
```
Should never be "null" or None

A-Artemis committed:
Refactored BaseWeightsService to create task directories on init.
Updated task from str to TaskType across objects.
Removed NullModelManifest in favour of raising exception when object not found.

A-Artemis committed:
Merge branch 'develop' of github.com:open-edge-platform/training_extensions into aurelien/base-weights-service

# Conflicts:
#	application/backend/app/api/dependencies.py

A-Artemis committed:
Raises ValueError to notify user that somthing went wrong during download

A-Artemis committed:
Fix test
